controllers/machine_controller_node.go

```diff
@@ -0,0 +1,150 @@
+				"cannot assign NodeRef to Machine %q in namespace %q, no matching Node", machine.Name, machine.Namespace)
+		}
+		logger.Error(err, "Failed to assign NodeRef")
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return err
+	}
+
+	// Set the Machine NodeRef.
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &apicorev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if err := hasHealthyNodeConditions(node); err != nil {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "Unhealthy conditions: %s", err.Error())
+		return nil
+	}
+
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+	return nil
+}
+
+// hasHealthyNodeConditions checks if a Node is healthy and does not have any semantically negative Conditions.
+// Returns error if NodeMemoryPressure, NodeDiskPressure, or NodePIDPressure Conditions are false or Ready Condition is true.
+func hasHealthyNodeConditions(node *apicorev1.Node) error {
+	errList := []error{}
+
+	for _, condition := range node.Status.Conditions {
+		switch condition.Type {
+		case apicorev1.NodeMemoryPressure, apicorev1.NodeDiskPressure, apicorev1.NodePIDPressure:
+			if condition.Status == apicorev1.ConditionTrue {
+				errList = append(errList, errors.Errorf("Condition %s is true on Node: %s", condition.Type, node.Name))
+			}
+		case apicorev1.NodeReady:
+			if condition.Status == apicorev1.ConditionFalse {
+				errList = append(errList, errors.Errorf("Condition %s is false on Node: %s", condition.Type, node.Name))
+			}
+		}
+	}
+	return kerrors.NewAggregate(errList)
+}
+
+func (r *MachineReconciler) getNode(c client.Reader, providerID *noderefutil.ProviderID) (*apicorev1.Node, error) {
+	logger := r.Log.WithValues("providerID", providerID)
+
+	nodeList := apicorev1.NodeList{}
+	for {
+		if err := c.List(context.TODO(), &nodeList, client.Continue(nodeList.Continue)); err != nil {
+			return nil, err
+		}
+
+		for _, node := range nodeList.Items {
+			nodeProviderID, err := noderefutil.NewProviderID(node.Spec.ProviderID)
+			if err != nil {
+				logger.Error(err, "Failed to parse ProviderID", "node", node.Name)
+				continue
+			}
+
+			if providerID.Equals(nodeProviderID) {
+				return &node, nil
+			}
+		}
+
+		if nodeList.Continue == "" {
+			break
+		}
+	}
+	return nil, ErrNodeNotFound
```

Comment:
wondering if `return nil, nil` is equally expressive to represent node not found (thus avoiding the need of `ErrorType`)